File: sdkwork-rtc-sdk-kotlin/src/main/kotlin/com/sdkwork/rtc/metadata/RtcProviderSupport.kt
package com.sdkwork.rtc.metadata

enum class RtcProviderSupportStatus {
    builtin_registered,
    official_registered,
    official_unregistered,
    unknown,
}

data class RtcProviderSupport(
    val providerKey: String,
    val status: RtcProviderSupportStatus,
    val builtin: Boolean,
    val official: Boolean,
    val registered: Boolean,
)

data class RtcProviderSupportStateRequest(
    val providerKey: String,
    val builtin: Boolean,
    val official: Boolean,
    val registered: Boolean,
)

val RTC_PROVIDER_SUPPORT_STATUSES: List<String> = listOf(
    "builtin_registered",
    "official_registered",
    "official_unregistered",
    "unknown",
)

fun resolveRtcProviderSupportStatus(
    request: RtcProviderSupportStateRequest,
): RtcProviderSupportStatus {
    if (request.official && request.registered) {
        return if (request.builtin) {
            RtcProviderSupportStatus.builtin_registered
        } else {
            RtcProviderSupportStatus.official_registered
        }
    }

    if (request.official) {
        return RtcProviderSupportStatus.official_unregistered
    }

    return RtcProviderSupportStatus.unknown
}

fun createRtcProviderSupportState(
    request: RtcProviderSupportStateRequest,
): RtcProviderSupport {
    return RtcProviderSupport(
        providerKey = request.providerKey,
        status = resolveRtcProviderSupportStatus(request),
        builtin = request.builtin,
        official = request.official,
        registered = request.registered,
    )
}

File: sdkwork-rtc-sdk-swift/Sources/RtcSdk/RtcProviderSupport.swift
public enum RtcProviderSupportStatus: String {
    case builtin_registered = "builtin_registered"
    case official_registered = "official_registered"
    case official_unregistered = "official_unregistered"
    case unknown = "unknown"
}

public struct RtcProviderSupportStateRequest {
    public let providerKey: String
    public let builtin: Bool
    public let official: Bool
    public let registered: Bool

    public init(
        providerKey: String,
        builtin: Bool,
        official: Bool,
        registered: Bool
    ) {
        self.providerKey = providerKey
        self.builtin = builtin
        self.official = official
        self.registered = registered
    }
}

public struct RtcProviderSupport {
    public let providerKey: String
    public let status: RtcProviderSupportStatus
    public let builtin: Bool
    public let official: Bool
    public let registered: Bool

    public static let rtcProviderSupportStatuses: [String] = [
        "builtin_registered",
        "official_registered",
        "official_unregistered",
        "unknown",
    ]

    public static func resolveRtcProviderSupportStatus(
        _ request: RtcProviderSupportStateRequest
    ) -> RtcProviderSupportStatus {
        if request.official && request.registered {
            return request.builtin ? .builtin_registered : .official_registered
        }

        if request.official {
            return .official_unregistered
        }

        return .unknown
    }

    public static func createRtcProviderSupportState(
        _ request: RtcProviderSupportStateRequest
    ) -> RtcProviderSupport {
        return RtcProviderSupport(
            providerKey: request.providerKey,
            status: resolveRtcProviderSupportStatus(request),
            builtin: request.builtin,
            official: request.official,
            registered: request.registered
        )
    }
}
